rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,393 @@
+    sha256: 'abcd1234',
+    size: 1024,
+  }, {
+    sha256: 'abcd1234',
+    size: 1024,
+  }],
+}
+```
+
+This endpoint will return a list of HTTP requests which must all be run.
+
+The `PATCH` endpoint will be sent without a request body and will be used to
+perform any post-upload actions.  These include things like the commit step on
+an S3 multipart upload.  This endpoint would be run by the frontend service and
+not by the uploader, and the frontend service would also need a complimentary
+method for workers to call when they complete the upload.  This is what is
+currently done with the `blob` storage type in the queue.
+
+The completion endpoint is used to perform any post-upload validation required.
+This might include the "Complete Multipart Upload" for S3 based objects, or
+possible a post-upload hash and verification step for a locally managed file
+server.
+
+Until the completion endpoint is called for a given object, attempts to
+retrieve an object will result in a 404 Not Found response.
+
+Object creation must be idempotent.  If a call to the creation end point
+happens with equivalent values, the same response body must be returned.  If a
+call is made to the endpoint with conflicting values, an error must be returned
+
+Example of a multipart upload using S3 from the object service's perspective.
+This is a simplified view, meant to highlight the key interactions rather than
+every detail.  For the purpose of this example, the uploader speaks directly
+with the object service.  In reality this interaction would have something like
+the Queue as an intermediary between the uploader and the object service.
+
+```
+uploader --> objsvc PUT /objects/abcd123  {
+                                            origin: '10.10.10.10',
+                                            contentType: 'application/json',
+                                            contentLength: 1234,
+                                            contentSha256: 'abcdef1234',
+                                            transferLength: 1234,
+                                            transferSha256: 'abcdef1234',
+                                            contentEncdoing: 
+                                            expiration: new Date(),
+                                            parts: [
+                                              sha256: 'abcd1234',
+                                              size: 1024,
+                                            }, {
+                                              sha256: 'abcd1234',
+                                              size: 1024,
+                                            }],
+                                          }
+objsvc --> s3 POST /object-abcd123?uploads
+objsvc <-- s3 200 UploadId=u-123
+uploader <-- objsvc 200 {requests: [{
+                          url: 'http://s3.com/object-abcd123?partNumber=1&uploadId=u-123',
+                          headers: {authorization: 'signature-request-1'},
```

Comment:
Consider echoing back out the metadata in the response.

Reply:
Happy to do that